src/components/KeyboardAvoidingView/index.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2015-present, Nicolas Gallagher.
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * @providesModule KeyboardAvoidingView
+ * @noflow
+ */
+
+import createReactClass from 'create-react-class';
+import StyleSheetPropType from '../../propTypes/StyleSheetPropType';
+import View from '../View';
+import ViewPropTypes from '../View/ViewPropTypes';
+import ViewStylePropTypes from '../View/ViewStylePropTypes';
+import React from 'react';
+import {number, oneOf} from 'prop-types';
+
+type ViewLayout = {
+  x: number,
+  y: number,
+  width: number,
+  height: number,
+};
+type ViewLayoutEvent = {
+  nativeEvent: {
+    layout: ViewLayout,
+  }
+};
+type ScreenRect = {
+  screenX: number,
+  screenY: number,
+  width: number,
+  height: number,
+};
+type KeyboardChangeEvent = {
+  startCoordinates?: ScreenRect,
+  endCoordinates: ScreenRect,
+  duration?: number,
+  easing?: string,
+};
+
+const viewRef = 'VIEW';
+
+/* eslint-disable react/prefer-es6-class, react/prop-types */
+const KeyboardAvoidingView = createReactClass({
+  propTypes: {
+    ...ViewPropTypes,
+    behavior: oneOf(['height', 'position', 'padding']),
+    contentContainerStyle: StyleSheetPropType(ViewStylePropTypes),
+    keyboardVerticalOffset: number.isRequired,
+    style: StyleSheetPropType(ViewStylePropTypes)
+  },
+
+  getDefaultProps() {
+    return {
+      keyboardVerticalOffset: 0,
+    };
+  },
+
+  frame: (null: ?ViewLayout),
+
+  relativeKeyboardHeight(keyboardFrame: ScreenRect): number {
```

Comment:
what's going on here?

Reply:
They keep a reference to the latest ViewLayout so the relativeKeyboardHeight method can compute the position of the current View